### PR TITLE
Add name propagation for at::alias, add tensor.set_names

### PR DESCRIPTION
**Pull request by zou3519**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #24087 Name inference for softmax, log_softmax and Dimname overloads.
* #24028 Fix out= function semantics for named tensors.
* #23973 Update tensor.view_names / tensor.names_ API
* #23962 Rename set_names -> view_names, set_names_ -> names_
* #23804 Implement tensor.align_to(names), torch.align_tensors(t1, t2, ...)
* #23808 Named inference for contiguous(), bernoulli variants, and dropout.
* #23746 Implement name inference rule for empty_like, clone
* #23743 Add `names` argument to ones, rand, randn, zeros, full; fix empty
* #23801 Improve test_namedtensor.py with named tensor equality check
* **#23624 Add name propagation for at::alias, add tensor.set_names**

tensor.set_names(names) is the out-of-place variant of
tensor.set_names_(names). This naming is probably confusing so I am
taking any and all suggestions.

Test Plan:
- run tests [namedtensor ci]

gh-metadata: pytorch pytorch 23624 gh/zou3519/86/head

Differential Revision: [D16621830](https://our.internmc.facebook.com/intern/diff/D16621830)

### CODE REVIEW

File: test/test_namedtensor.py [hunk 1]
```diff
@@ -115,6 +115,20 @@ def test_set_names_(self):
...
+        tensor = torch.empty(1, 1, names=('N', 'C'))
+
+        self.assertEqual(tensor.set_names(None).names, (None, None))
+        self.assertEqual(tensor.set_names(['H', 'W']).names, ('H', 'W'))
+
+        # Check that we didn't modify tensor.names
+        self.assertEqual(tensor.names, ('N', 'C'))
+
+        with self.assertRaisesRegex(RuntimeError, 'Number of names'):
+            tensor.set_names(['N', 'C', 'W'])
```
Comment (zou3519) [inline test/test_namedtensor.py]: @nairbv I remember you had a concern about the naming of `tensor.set_names_`. What do you think about changing `set_names` and `set_names_` to `tensor.name_dims` and `tensor.name_dims_`?

Comment (nairbv) [reply test/test_namedtensor.py]: I think I'd prefer `dim_names`/`dim_names_` over `name_dims`, just because the later sounds like 'name' is being used as a verb even in the get case.

Comment (zou3519) [reply test/test_namedtensor.py]: > I think I'd prefer dim_names/dim_names_ over name_dims, just because the later sounds like 'name' is being used as a verb even in the get case.

I agree that some noun would be better. `dim_names` sounds too much like the `names` attribute. Maybe something like `tensor.updated_names / tensor.updated_names_`.

Aside from the naming issue, are there any other points in this PR that you had concerns over? I'd prefer to punt a full resolution for the names to a future PR because I need an out-out-place name setter to implement things further down the stack.

**Review by nairbv — APPROVED** [review]

File: test/test_namedtensor.py (hunk 1 above)
Comment (nairbv) [reply test/test_namedtensor.py]: what's the `names` attribute? I thought that's what this is setting?

Comment (zou3519) [reply test/test_namedtensor.py]: > what's the names attribute? I thought that's what this is setting?

Yes, that's what this is setting. I was just worried that `dim_names` sounds too much like an attribute and less like a method and that users might confuse it with `names.`

`tensor.names = ('N', 'C', 'H', 'W')` is another way of writing `tensor.set_names_(['N', 'C', 'H', 'W'])`

Comment (nairbv) [reply test/test_namedtensor.py]: this is in python? can the user just `tensor.names` to get the names? if so maybe the setter should then just be `tensor.names_()` to have them all be the same?

Comment (zou3519) [reply test/test_namedtensor.py]: > this is in python?

Yes

>  can the user just tensor.names to get the names?

Yes, that is the intention

> if so maybe the setter should then just be tensor.names_() to have them all be the same?

That's reasonable. The one concern I have with that is that usually an out-of-place version of a method is just the method without an underscore; we can't have it be called `tensor.names()` because `names` is an attribute.

Comment (nairbv) [reply test/test_namedtensor.py]: Trying to think what's most consistent with what we have now... 

Would we consider this a tensor attribute? (https://pytorch.org/docs/stable/tensor_attributes.html)
For those we use `tensor.to(newDtype)` or tensor.to(newDevice)`. There is no `tensor.dtype_` or `tensor.set_dtype()` ... but those return a new tensor instead of mutating the existing one right?

`requires_grad` / `requires_grad_()` seem similar to the `names` attribute, but not sure if that's the best example.

Comment (zou3519) [reply test/test_namedtensor.py]: I don't think this should be done using a `.to(...)` function because those are for "heavier" changes that involve transforming tensor data.

However, I agree that it feels more like requires_grad. We have a `requires_grad` attribute and `requires_grad_()` mutator, but we don't really have an out-of-place `requires_grad` method. (we do have the "create-a-new-tensor-from-your-old-tensor-with-requires_grad=True" option).

Comment (zou3519) [reply test/test_namedtensor.py]: >  but we don't really have an out-of-place requires_grad method. 

To drill down further into this point, I think the reason why it doesn't exist is because it's not clear what it should do:
1) return a new tensor with copied data with requires_grad=True
2) return a view of the old tensor with requires_grad=True
